strawberryfields/backends/fockbackend/circuit.py

```diff
@@ -314,7 +315,14 @@ def beamsplitter(self, t, r, phi, mode1, mode2):
         """
         Applies a beamsplitter.
         """
-        self._apply_gate(ops.beamsplitter(t, r, phi, self._trunc), [mode1, mode2])
+        mat = ops.beamsplitter(t, r, phi, self._trunc)
+
+        modes = List()
+        modes.append(mode1)
+        modes.append(mode2)
+
+        args = [mat, self._state, self._pure, modes, self._num_modes, self._trunc]
+        self._state = ops.apply_twomode_gate(*args, gate="BSgate")
```

Comment:
Based on the arguments being passed to `apply_twomode_gate`, it sounds like it should be a method in the `Circuit` class 🙂 This way, you won't need to pass `self._state`, `self._pure`, `self._num_modes`, `self._trunc`

Reply:
I agree about that, although I put the `apply_twomode_gate` in ops.py since all other tensor contraction functions are there (`apply_gate_blas` and `apply_gate_einsum`). All three have the same arguments. I'm not sure if separating them is a good idea.

Reply:
Another option would be to put all tensor contractions in the `Circuit` class, since they seem to only be called from within there. I don't know if that would brake something else though. :slightly_smiling_face: